File: skin_care_agent/backend/app/domain/region_catalog.py
from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import Literal, cast


RegionId = Literal[
    "forehead",
    "left_face",
    "right_face",
    "nose_area",
    "mouth_area",
    "chin",
]

REGION_IDS: tuple[RegionId, ...] = (
    "forehead",
    "left_face",
    "right_face",
    "nose_area",
    "mouth_area",
    "chin",
)


@dataclass(frozen=True)
class RegionDefinition:
    region_id: RegionId
    label: str
    direction_note: str


REGION_DEFINITIONS: dict[RegionId, RegionDefinition] = {
    "forehead": RegionDefinition("forehead", "额头", "面部中央上方的额头区域"),
    "left_face": RegionDefinition(
        "left_face",
        "你的左侧脸",
        "用户本人真实左侧；自拍预览是否镜像都不改变这个方向",
    ),
    "right_face": RegionDefinition(
        "right_face",
        "你的右侧脸",
        "用户本人真实右侧；自拍预览是否镜像都不改变这个方向",
    ),
    "nose_area": RegionDefinition("nose_area", "鼻周", "鼻部及紧邻鼻翼范围"),
    "mouth_area": RegionDefinition("mouth_area", "口周", "嘴唇、嘴角及周围范围"),
    "chin": RegionDefinition("chin", "下巴", "下唇下方至下巴下缘中央"),
}


def normalize_region_ids(values: Iterable[str]) -> tuple[RegionId, ...]:
    received = tuple(values)
    if not 1 <= len(received) <= len(REGION_IDS):
        raise ValueError("select one to six regions")
    if len(set(received)) != len(received):
        raise ValueError("region IDs must be unique")
    if any(value not in REGION_IDS for value in received):
        raise ValueError("unsupported region ID")
    selected = set(received)
    return tuple(cast(RegionId, region_id) for region_id in REGION_IDS if region_id in selected)

File: skin_care_agent/mobile/src/lib/region-catalog.ts
export const REGION_IDS = [
  'forehead',
  'left_face',
  'right_face',
  'nose_area',
  'mouth_area',
  'chin',
] as const;

export type RegionId = (typeof REGION_IDS)[number];

export type RegionDefinition = {
  id: RegionId;
  label: string;
  boundary: string;
};

export const REGIONS: readonly RegionDefinition[] = [
  { id: 'forehead', label: '额头', boundary: '面部中央上方的额头区域' },
  {
    id: 'left_face',
    label: '左脸颊',
    boundary: '用户本人真实左侧；自拍预览是否镜像都不改变这个方向',
  },
  {
    id: 'right_face',
    label: '右脸颊',
    boundary: '用户本人真实右侧；自拍预览是否镜像都不改变这个方向',
  },
  { id: 'nose_area', label: '鼻周', boundary: '鼻部及紧邻鼻翼范围' },
  { id: 'mouth_area', label: '口周', boundary: '嘴唇、嘴角及周围范围' },
  { id: 'chin', label: '下巴', boundary: '下唇下方至下巴下缘中央' },
] as const;

const REGION_ID_SET: ReadonlySet<string> = new Set(REGION_IDS);

export function isRegionId(value: unknown): value is RegionId {
  return typeof value === 'string' && REGION_ID_SET.has(value);
}

export function normalizeRegionIds(values: readonly unknown[]): RegionId[] {
  if (values.length < 1 || values.length > REGION_IDS.length) {
    throw new Error('请选择一到六个区域。');
  }
  if (!values.every(isRegionId)) {
    throw new Error('包含不支持的区域。');
  }
  if (new Set(values).size !== values.length) {
    throw new Error('区域不能重复。');
  }
  const selected = new Set(values);
  return REGION_IDS.filter((regionId) => selected.has(regionId));
}

export function regionById(regionId: RegionId): RegionDefinition {
  return REGIONS.find(({ id }) => id === regionId)!;
}
